File: testsuite/tests/ada/0016-access-to-subp/cpp/main.cpp
#include <cstdint>
#include <iostream>

#include "test.h"

using namespace gnatpolyglot::ada::arrays;

class Child : public test::T {
public:
    Child() : test::T (this) {}

    int32_t t_call(std::function<int32_t (int32_t, int32_t)> c) const override {
        return c(-1, -2);
    }
};

std::int32_t foo(std::int32_t i, std::int32_t j) {
  std::cout << "Hello!\n";
  return i + j;
}

int main() {
    {
        // Scalar callbacks
        std::cout << test::call(&foo) << "\n";
        std::cout << test::call([](int i, int j) { return i * j; }) << "\n";
        // Record callbacks
        test::Rec r = test::call([](test::Rec rec) {
            return test::Rec(rec.get_i() - 1, rec.get_j() - 2);
        });
        std::cout << "Rec{" << r.get_i() << ", " << r.get_j() << "}\n";
        // Array callbacks
        polyglot_array<test::Rec> arr = test::call([](const polyglot_array<test::Rec> &arr) {
            polyglot_array<test::Rec> res = arr;
            for (auto &rec : res) {
                rec.get_i() += 4;
                rec.get_j() += 8;
            }
            return res;
        });
        std::cout << "{" << "Rec{" << arr.get(1)->get_i() << ", " << arr.get(1)->get_j() << "}, "
                  << "{" << "Rec{" << arr.get(2)->get_i() << ", " << arr.get(2)->get_j() << "}}\n";
    }

    std::cout << std::endl;

    {
        // Scalar callbacks
        std::cout << test::get_callback_1()(10, 8) << "\n";
        // Record callbacks
        test::Rec r = test::get_callback_2()(test::Rec(10, 8));
        std::cout << "Rec{" << r.get_i() << ", " << r.get_j() << "}\n";
        // Array callbacks
        polyglot_array<test::Rec> arr1{1, 2};
        arr1.set(1, test::Rec{1, 2});
        arr1.set(2, test::Rec{3, 4});
        polyglot_array<test::Rec> arr2 = test::get_callback_3()(arr1);
        std::cout << "{" << "Rec{" << arr2.get(1)->get_i() << ", " << arr2.get(1)->get_j() << "}, "
                  << "{" << "Rec{" << arr2.get(2)->get_i() << ", " << arr2.get(2)->get_j() << "}}\n";
    }

    std::cout << std::endl;

    std::function<int32_t(int32_t, int32_t)> f = nullptr;
    test::get_callback(f);
    std::cout << f(10, 8) << "\n";

    test::T t{};
    Child c{};
    std::cout << test::call_t_call(t, &foo) << "\n";
    std::cout << test::call_t_call(c, &foo) << "\n";

}

File: testsuite/tests/ada/0016-access-to-subp/java/Main.java
import com.adacore.libtest.test.TestPackage;
import com.adacore.gnatpolyglot.runtime.Functions.Function2;
import com.adacore.libtest.test.T;
import com.adacore.libtest.test.Rec;

public class Main {

    static class Child extends T {
        @Override
        public int tCall(com.adacore.libtest.Callbacks.Callback0 c) {
            return c.apply(-1, -2);
        }
    }

    static int foo(int i, int j) {
        System.out.println("Hello!");
        return i + j;
    }

    public static void main(String[] args) throws Throwable {
        {
            // Scalar callbacks
            System.out.println(TestPackage.call(Main::foo));
            System.out.println(TestPackage.call((i, j) -> i * j));
            // Record callbacks
            Rec r = TestPackage.call(
                (Rec rec) -> new Rec(rec.getI() - 1, rec.getJ() - 2)
            );
            System.out.println("Rec{%d, %d}".formatted(r.getI(), r.getJ()));
            // Array callbacks
            Rec.Array arr = TestPackage.call(
                (Rec.Array a) -> {
                    Rec.Array res = new Rec.Array(a);
                    for (var rec : res) {
                        rec.setI(rec.getI() + 4);
                        rec.setJ(rec.getJ() + 8);
                    }
                    return res;
                }
            );
            System.out.println(
                "{Rec{%d, %d}, Rec{%d, %d}}".formatted(
                    arr.get(0).getI(), arr.get(0).getJ(),
                    arr.get(1).getI(), arr.get(1).getJ()
                )
            );
        }

        System.out.println();

        {
            // Scalar callbacks
            System.out.println(TestPackage.getCallback1().apply(10, 8));
            // Record callbacks
            Rec r = TestPackage.getCallback2().apply(new Rec(10, 8));
            System.out.println("Rec{%d, %d}".formatted(r.getI(), r.getJ()));
            // Array callbacks
            Rec.Array arr1 = new Rec.Array(1, 2);
            arr1.set(0, new Rec(1, 2));
            arr1.set(1, new Rec(3, 4));
            Rec.Array arr2 = TestPackage.getCallback3().apply(arr1);
            System.out.println(
                "{Rec{%d, %d}, Rec{%d, %d}}".formatted(
                    arr2.get(0).getI(), arr2.get(0).getJ(),
                    arr2.get(1).getI(), arr2.get(1).getJ()
                )
            );
        }

        System.out.println();

        com.adacore.libtest.Callbacks.Callback0.Ref f =
            new com.adacore.libtest.Callbacks.Callback0.Ref();
        TestPackage.getCallback(f);
        System.out.println(f.get().apply(10, 8));

        T t = new T();
        Child c = new Child();
        System.out.println(TestPackage.callTCall(t, Main::foo));
        System.out.println(TestPackage.callTCall(c, Main::foo));
    }
}
